### ISSUE TITLE
`IStrategy` should accept multiple input types with default implementations

### ISSUE THREAD

**Issue by mmmcaffeine** [mmmcaffeine/nonograms#9]
Assuming we have types of `Line` (#1) and `Hint` (#2) the `IStrategy` interface will change to accept these types, rather than arrays of primitives as it currently does. There will almost certainly be conversion operators or `Parse` implementations to e.g. convert a `string` to a `Hint`. It would make sense for the `IStrategy` interface to have overloads that accept these types with a default implementation of e.g.

`public Line Execute(string hint, string line) => Execute(Hint.Parse(hint), Line.Parse(line));`

This _might_ be an abuse of the typical use case for this language feature (add new methods to an existing interface) so we _might_ be better off putting the implementation on `StrategyBase`. However, I cannot see there would ever be any other implementation of such methods, so it existing on the interface eases life for implementers who _don't_ want to inherit `StrategyBase`.

**Comment by mmmcaffeine** [comment]
It turns out we _don't_ need overloads that accept the arrays. That's fairly obvious really. Because the `Hint` and `Line` types implement implicit conversions we can simply pass them into the original signature.

Although we have implicit conversions _from_ `Hint` and `Line` _to_ `string` we quite deliberately _don't_ have them in the other direction; the likelihood of the value not being a valid representation is high and we can't (shouldn't) have exceptions from implicit conversions. As such we _do_ still need to define an overload that accepts strings and delegates through to the respective `Parse` methods.

**Comment by mmmcaffeine** [comment]
The last time I tried to use [FakeItEasy](https://github.com/FakeItEasy/FakeItEasy) with default interface implementations was in my [Minesweeper repo](https://github.com/mmmcaffeine/minesweeper). I found things a bit clumsy, and was quite disappointed with the outcome 😞

I didn't like the name of the test fixture; it sound like an interface! I also found the fakes did _not_ pick up the implementations defined on the interface itself. In practice that meant I ended up having to mock a _lot_ of things I didn't expect to. Tests of types that consumed the interface with default implementations became very hard to read. I eventually decided to dump the fakes, and write integration tests using real types.

With a bit more thought and experimenting I have found a much more acceptable approach here. The name for test fixture doesn't follow the usual convention of `{SutType}Tests`, but no longer looks like an interface. Creating an abstract type that implements the interface I am attempting to test resolves the problem of not picking up default implementations. The abstract type (rather than the interface) can then be faked. This means we can still define behaviour on the interface methods that don't have default implementations, and assert against it, but still get the default implementations! 😀
